codegen/src/main/java/software/amazon/awssdk/codegen/poet/authscheme/AuthSchemeParamsSpec.java

```diff
@@ -0,0 +1,131 @@
+
+        addBuilderSetterMethods(b);
+
+        b.addMethod(MethodSpec.methodBuilder("build")
+                              .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+                              .returns(className())
+                              .addJavadoc("Returns a {@link $T} object that is created from the properties that have been set "
+                                          + "on the builder.", authSchemeSpecUtils.parametersInterfaceName())
+                              .build());
+
+        return b.build();
+    }
+
+    private void addAccessorMethods(TypeSpec.Builder b) {
+        b.addMethod(MethodSpec.methodBuilder("operation")
+                              .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+                              .returns(String.class)
+                              .addJavadoc("Returns the operation for which to resolve the auth scheme.")
+                              .build());
+
+        if (authSchemeSpecUtils.usesSigV4()) {
+            b.addMethod(MethodSpec.methodBuilder("region")
+                                  .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+                                  // TODO: reviewer: Should region be Regions (Regions.class isn't available here though)
+                                  .returns(ParameterizedTypeName.get(Optional.class, String.class))
+                                  .addJavadoc("Returns the region. The region is optional. The region parameter may be used "
+                                              + "with $S auth scheme. By default, the region will be empty.", "aws.auth#sigv4")
+                                  .build());
+        }
+    }
+
+    private void addBuilderSetterMethods(TypeSpec.Builder b) {
+        b.addMethod(MethodSpec.methodBuilder("operation")
+                              .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+                              .addParameter(ParameterSpec.builder(String.class, "operation").build())
+                              .returns(authSchemeSpecUtils.parametersInterfaceBuilderInterfaceName())
+                              .addJavadoc("Set the operation for which to resolve the auth scheme.")
+                              .build());
+
+        if (authSchemeSpecUtils.usesSigV4()) {
+            b.addMethod(MethodSpec.methodBuilder("region")
+                                  .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+                                  .addParameter(ParameterSpec.builder(String.class, "region").build())
+                                  .returns(authSchemeSpecUtils.parametersInterfaceBuilderInterfaceName())
+                                  .addJavadoc("Set the region. The region parameter may be used with  the $S auth scheme.",
+                                              "aws.auth#sigv4") // TODO: Could the known auth schemes be defined somewhere?
```

Comment:
If we're not exposing the auth scheme id (slightly unsure), do we need to reference it? Can we reference the AuthScheme class for Sigv4 instead?

Reply:
I think they can be considered public. I used it in the javadoc for [HttpAuthScheme](https://github.com/aws/aws-sdk-java-v2/blob/28ddc6018daf96b8c55e2e8137fca2f86c1d42c4/core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/HttpAuthScheme.java#L41). Good point about using the AuthScheme class. It doesn't exist yet. So will keep the TODO to update when that exists.